Refactor the upsell flow to wait for decline button to be present

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

- refactor `UpsellPage.declineOffer` function to use named CSS selector and add waits to guard.
- update test step `Can search for a blog name` to search for the `blogName` + .com TLD instead of just the `blogName` as the .com TLD which is used in the test is not always the default suggestion.

#### Testing instructions

Run tests on CircleCI.

Check for the following test:
`Adding a domain to an existing site `
`Calypso Gutenberg Editor: Checkout on` - subtest accesses the upsell page.

worldomonation committed:
Add step to findADomainComponent.selectDomainAddress() to click on the .com TLD button to ensure tests do not fail unintentionally when .com domain is not in the suggestions.

worldomonation committed:
Scroll the decline button into view prior to clicking it.

worldomonation committed:
Additional tweaks to findADomainComponent.selectDomainAddress() to wait until the .com TLD button is visible before attempting to click.

worldomonation committed:
Correct selector for findADomainComponent.selectDomainAddress().

worldomonation committed:
Add additional wait for element to be present and displayed when publishing post in GutenbergEditorComponent.

worldomonation committed:
Remove action to click on .com TLD button from findADomainComponent due to issues with mobile.
Instead, update the test to search explicitly for the .com TLD domain.

worldomonation committed:
Wait until upsell cancellation button is visible before scrolling into view and clicking.

worldomonation committed:
Do not require closure of panel when publishing post in wp-calypso-gutenberg-post-editor-spec.js.
Additional sleep when navigating away from checkout page in wp-manage-domains-spec.js.

worldomonation committed:
Extraneous new line.

worldomonation committed:
Try out additional sleeps in the GutenbergEditorComponent.publish.

worldomonation committed:
Refactor some functions in GutenbergEditorComponent.
- clicking on buttons and elements are now done with `clickWhenClickable` function call where possible.
- moved declaration for CSS selector for snackbar down closer to where it is used.

worldomonation committed:
Fix the test behavior so that it will pass for the time being.
- set the `closePanel` argument to false for `Can publish and view content` test step due to issues with Payments with Paypal block.
- filed issue https://github.com/Automattic/wp-calypso/issues/50302.

worldomonation committed:
Update Insert a Click to Tweet block test.
- remove the closePanel:false parameter as the panel should be closed to access the View Post snackbar, especially on mobile.

bsessions85 commented:
@worldomonation  You'll want to remove the chromedriver log files in there.

worldomonation commented:
> @worldomonation You'll want to remove the chromedriver log files in there.

Oh dear, I can't believe I did that. Fixing.

worldomonation committed:
Remove accidental log file commit.

worldomonation committed:
Restore the sleep statement that was removed by accident.
Add a comment explaining the scrollIntoView.

worldomonation commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -81,12 +80,14 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
 		if ( visit ) {
-			const snackbar = await this.driver.findElement( snackBarNoticeLinkSelector );
-			await this.driver.executeScript( 'arguments[0].click();', snackbar );
+			await driverHelper.clickWhenClickable( this.driver, snackBarNoticeLinkSelector );
```
Instead of clicking the button using a script, let's take into account whether it's ready and rendered on screen first.

bsessions85 replied:
This is worth a shot. We changed it a while back because clickWhenClickable wasn't working.

worldomonation commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -523,10 +524,10 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
 	async closePublishedPanel() {
-		const closeButton = await this.driver.findElement(
+		return await driverHelper.clickWhenClickable(
+			this.driver,
```
Same idea with this change as above.

worldomonation commented on test/e2e/specs/wp-calypso-gutenberg-page-editor-spec.js:
```diff
@@ -555,7 +555,9 @@ describe( `[${ host }] Calypso Gutenberg Editor: Pages (${ screenSize })`, funct
-			return await gEditorComponent.publish( { visit: true } );
+			// closePanel is set to false because the panel is forcibly dismissed after publishing.
+			// See https://github.com/Automattic/wp-calypso/issues/50302.
+			return await gEditorComponent.publish( { closePanel: false, visit: true } );
```
Issue is on file for unexpected behavior when certain blocks are present in the editor and the post/page is published.

worldomonation committed:
Reduce scope of the PR to deal with just the upsell page.
- wait until the upsell decline button is visible on page prior to proceeding.
- update the test `Can search for a blog name` to use the full blog name including the `.com` TLD to reduce instances where different suggestions cause the test failure.

worldomonation commented:
Calypso tests passed.

Desktop e2e test has failure with the launch flow - which is addressed in #50411, so for purpose of this PR it can be ignored as it is a known flakey test.

bsessions85 reviewed (approved):
:shipit:
